compiler/transforms/test/debug_to_func.py

```diff
@@ -0,0 +1,71 @@
+from xdsl.context import MLContext
+from xdsl.dialects import arith, builtin, func, memref
+from xdsl.irdl.operations import Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.traits import SymbolTable
+
+from compiler.dialects.test import debug
+
+
+class DebugToFunc(RewritePattern):
+    """Insert debugging function calls"""
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: debug.DebugOp, rewriter: PatternRewriter):
+        ptr_a = memref.ExtractAlignedPointerAsIndexOp.get(op.op_a)
+        ptr_b = memref.ExtractAlignedPointerAsIndexOp.get(op.op_b)
+        ptr_c = memref.ExtractAlignedPointerAsIndexOp.get(op.op_c)
+
+        ops_to_insert: list[Operation] = [ptr_a, ptr_b, ptr_c]
+
+        ptr_a = arith.IndexCastOp(ptr_a, builtin.i32)
+        ptr_b = arith.IndexCastOp(ptr_b, builtin.i32)
+        ptr_c = arith.IndexCastOp(ptr_c, builtin.i32)
+
+        ops_to_insert.extend([ptr_a, ptr_b, ptr_c])
+
+        match (op.when.data, op.level.data):
+            case "before", "L3":
+                whenparam = 0
+            case "before", "L1":
+                whenparam = 1
+            case "after", "L1":
+                whenparam = 2
+            case "after", "L3":
+                whenparam = 3
+            case _:
+                whenparam = 5
+
+        when = arith.Constant.from_int_and_width(whenparam, 32)
+        ops_to_insert.append(when)
+
+        func_call = func.Call(
+            f"debug_{op.debug_type.data}", [ptr_a, ptr_b, ptr_c, when], []
+        )
+        ops_to_insert.append(func_call)
+        rewriter.replace_matched_op(ops_to_insert)
+
+        func_decl = func.FuncOp.external(
+            f"debug_{op.debug_type.data}",
+            [builtin.i32, builtin.i32, builtin.i32, builtin.i32],
+            [],
+        )
```

Comment:
Are these runtime thingies also added to this PR?

Reply:
no, the idea is you can specify your own in `main.c` and inspect anything you want in C code

Reply:
Oh okay cool! Can we at least have one test for this mechanism?

Reply:
The streamer-alu kernel uses this mechanism, and outputs the nice code:
```
Debugging at t = 0 with A at 80002F28, B at 80002FA8, C at 80003028
i0 -> A=44, B=39, C=0
i1 -> A=47, B=87, C=0
i2 -> A=64, B=46, C=0
i3 -> A=67, B=88, C=0
i4 -> A=67, B=81, C=0
Debugging at t = 1 with A at 10000040, B at 10000100, C at 100001C0
i0 -> A=44, B=39, C=0
i1 -> A=47, B=87, C=0
i2 -> A=64, B=46, C=0
i3 -> A=67, B=88, C=0
i4 -> A=67, B=81, C=0
Debugging at t = 2 with A at 10000040, B at 10000100, C at 100001C0
i0 -> A=44, B=39, C=83
i1 -> A=47, B=87, C=134
i2 -> A=64, B=46, C=110
i3 -> A=67, B=88, C=155
i4 -> A=67, B=81, C=148
Debugging at t = 3 with A at 80002F28, B at 80002FA8, C at 80003028
i0 -> A=44, B=39, C=83
i1 -> A=47, B=87, C=134
i2 -> A=64, B=46, C=110
i3 -> A=67, B=88, C=155
i4 -> A=67, B=81, C=148
```

Reply:
this prints before copying in L3, after copying in L1, after compute in L1 and then after copying back to L3